### DISCUSSION TITLE
Swift as a Godot Scripting Language

### DISCUSSION THREAD

**Discussion by atlasapplications** [migueldeicaza/SwiftGodot#515] [General]
It seems that in `SwiftGodot`'s current state it could be viable to use `Swift` as a scripting language, but compared to `C#` and `GDScript` the workflow is slower and a bit more convoluted (especially when you're not using Mac) since this plugin technically isn't implemented as a scripting language. 

If you look at `GDScript` and `C#` these are scripting languages because you can create a `Node` and attach one of these scripts to the `Node` to add functionality. However, for `SwiftGodot`'s implementation, it works closer to a `C++` module where you're actually extending the engine and that newly created `Node` is what's added to the scene. 

This approach is great when adding native functionality, which seems to be the focus of `GDExtension`. However, as a scripting language, there's an unnecessary loss in efficiency since this workflow requires more steps. One way I think of it is that `GDExtension` enables the creation of native plugins in `C++` and `GodotSwift` enables the creation of native plugins in `Swift`.

So, I'm wondering if there's any desire for Swift being implemented as a scripting language? 

An excellent example of `GDExtension` being used to create a new scripting language is Godot [Orchestrator](https://github.com/CraterCrash/godot-orchestrator). Perhaps this could be considered an enhancement and `Swift` would become a proper scripting language and still provide an avenue for extending the engine? This appears to be on the [roadmap](https://github.com/godotengine/godot-proposals/issues/7895)  for `C#` but in reverse with it currently being a scripting language and adding the ability for creating plugins using `GDExtension` in a future Godot version.

**Comment by Qws** [comment]
Swift as an (official) scripting language in Godot, same level as C# or even GDscript, would be so great.

**Comment by JacobSyndeo** [comment]
+1 for this. I really hope we can see it happen someday; I'd love that!

**Comment by GuyWithBag** [comment]
+1 since I think swift is the most similar to GDScript with typing features

**Comment by migueldeicaza** [comment]
I really need to sit down and re-bootstrap this effort.

I think the big thing that stopped me last time was that I really wanted to have the Swift plugin be "global" rather than be local to the project, so I did not have to struggle between the Swift plugin and the generated plugin by the user.

**Comment by migueldeicaza** [comment]
Btw, for those curious, the old branch that does this is on 

https://github.com/migueldeicaza/SwiftGodot/tree/editor-plugin

**Comment by YamikoHikariX** [comment]
Hi! I am really interested in using Swift to develop games in Godot on Windows. I honestly can't fully understand yet if this project is meant for this purpose but if I understand correctly it's not? Seems like it's mostly for extensions or something like that. This post seems to be the most relevant. Are there any updates on this? Can anyone please say how possible is it to use SwiftGodot to develop games on WIndows?

**Comment by migueldeicaza** [comment]
Yes, you can use swift to develop games with Godot on windows.

the discussion is about more closely integrating into the Godot editor

**Reply by YamikoHikariX** [reply]
Thanks! Seems like I got it to work but I guess won't use it for now. Unfortunately a bit too much hassle to rebuild and restart godot to see changes. Maybe will try at some point in the future again. Nice to know that it's even possible though. Like many others I'd also definitely be interested in seeing it become more integrated! I myself unfortunately have no idea how to contribute to that

**Comment by migueldeicaza** [comment]
Btw, I suspect this could be vibe coded at this point.

the reason it never got done, is because it was straightforward but tedious

**Comment by lexum0** [comment]
Any news on using Swift for scripting? 

It would be great!

**Comment by migueldeicaza** [comment]
Not yet, busy with the upcoming GodotCon :-)

**Comment by lexum0** [comment]
I don't have enough knowledge to implement Swift as a scripting language, if I were to do it would you tell me how?

**Comment by migueldeicaza** [comment]
Well, it would take me time to explain and research to give you pointers, it is something that could be navigated in the community on the Discord server, so you are not waiting on me.   But the basic idea is to hook up to the existing set of APIs that Godot exposes for this purpose and provide the implementation in an add-on.
